torch/distributed/nn/api/remote_module.py

```diff
@@ -241,6 +278,30 @@ def get_module_rref(self) -> rpc.RRef:
         """
         return self.module_rref
 
+    def __getstate__(self):
+        # Only pickle the attributes in _REMOTE_MODULE_PICKLED_ATTRIBUTES,
+        # and check if unpickled attributes are all in _REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING.
+        attrs = {}
+        for k, v in self.__dict__.items():
+            if k in _REMOTE_MODULE_PICKLED_ATTRIBUTES:
+                attrs[k] = v
+            elif k not in _REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING:
+                raise RuntimeError(
+                    "Attribute ``{}`` of RemoteModule must be either in ``_REMOTE_MODULE_PICKLED_ATTRIBUTES`` "
+                    " or ``_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING``.".format(k)
+                )
+
+        return attrs
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+        # Install generated methods when unpickled.
```

Comment:
Does this mean the RemoteModule on the remote end will get the forward_async generated method when it is unpickled on the other end? Does generating these methods during unpickling incur any significant overhead?

Reply:
> Does this mean the RemoteModule on the remote end will get the forward_async generated method when it is unpickled on the other end?

Yes, only if we pass a remote module as an arg in RPC.

> Does generating these methods during unpickling incur any significant overhead?

The overhead really depends on the cost of running `torch.jit.export` on both `forward` and `forward_async`. I don't have a good answer here. We have already paid this cost and installed these methods on the worker that creates a remote module.

Note that here the scenario is that, if we want to pass this remote module to a new node and let that node initiate forward/forward_async, then we have to do this when unpickling the remote module on the new node. This overhead is not avoidable if the user wants to pass a remote module to an RPC call.